_posts/24-07-2022-what_is_a_reader_and_why_it_matters.md

```diff
@@ -0,0 +1,134 @@
+
+```Scala
+def i: A => D = f andThen g andThen h
+```
+
+### The reality
+This kind of composition, obviously, works only for functions with one argument - with multi-argument functions we quickly run into a world of trouble. Quite unfortunate, since the vast majority of those we encounter on a daily basis need more than one input. But what if all of our functions look like this?:
+
+```Scala
+def f: E => A => B = ???
+def g: E => B => C = ???
+def h: E => C => D = ???
+```
+
+This is again a common thing - oftentimes we need to pass some sort of a context (or other dependency) to our method. It would be useful to be able to compose them, even if just to save us some writing.
+They share the type of the first argument - it must have made our task easier. What do we need to do to be able to compose them in a sensible way?
+
+## The promise
+First of all, let's simplify things a bit and imagine a function that magically reads some value `V` from an environment `E`. It's going to have a type `E => V`. We're going to call it `Reader` for precisely that reason, and we're going to wrap it in a case class for convenience:
+
+```Scala
+case class Reader[E, V](read: E => V)
+```
+
+We're halfway there, we just need to notice that function parameters should be swappable (with possible minor tweaks to the logic inside - but usually we don't depend on their order anyway). Then we can rewrite our `f`, `g`, and `h` as:
+
+```Scala
+def f: A => E => B = ???
+def g: B => E => C = ???
+def h: C => E => D = ???
+```
+
+There's a lot of `Reader`s just waiting in line. We can now rewrite our functions again:
+
+```Scala
+def f: A => Reader[E, B] = ???
+def g: B => Reader[E, C] = ???
+def h: C => Reader[E, D] = ???
+```
+
+### The solution
+Now we just need to compose these somehow. First of all, let's give our Reader a helper method:
+
+```Scala
+case class Reader[E, V](read: E => V) {
+  def pipeTo[W](reader: V => Reader[E, W]): Reader[E, W] = Reader { env =>
+    val result1 = read(env)
+    reader(result1).read(env)
+  }
+}
+```
+
+Then we can make ourselves a helper:
+
+```Scala
+object Reader {
+  implicit class ReaderOps[E, A, B](f1: A => Reader[E, B]) {
+    def andThenR[C](f2: B => Reader[E, C]): A => Reader[E, C] =
+      a => f1(a).pipeTo(f2)
+  }
+}
+```
+
+Et voilà, we can now write this:
+
+```Scala
+def i: A => Reader[E, D] = f andThenR g andThenR h
+```
+
+And it works!
+
+### A bit of black magic
+We can do one more trick. Let's first refactor the Reader to be able to use Scala's wonderful for-comprehensions by renaming `pipeTo` to `flatMap` and adding a `map` method:
+
+```Scala
+case class Reader[E, V](read: E => V) {
+  def map[W](f: V => W): Reader[E, W] = Reader(env => f(read(env)))
+  def flatMap[W](f: V => Reader[E, W]): Reader[E, W] = Reader { env =>
+    val result1 = read(env)
+    f(result1).read(env)
+  }
+}
+```
+
+and then define:
+
+```Scala
+def ask[E]: Reader[E, E] = Reader(identity)
```

Comment:
Probably the name has some obscure Haskell origin? ;) doesn't sound well-chosen TBH

Reply:
Do Scala libs like Cats use the same naming convention

Reply:
Interestingly enough, `Reader` from Cats doesn't have an `ask` function like this. However in Cats `Reader[-E, A]` is an alias for `ReaderT[Id, E, A]`, and `ReaderT[F[_], -E, A]` is in turn an alias for `Kleisli[F, A, B]`, which defines an `ask[F[_], A]`. A bit more convoluted, but it can be used in similar fashion to what's implemented here - you just need to call it like this: `ask[Id, WhateverEnvYouHave]`.

Name most likely comes from Haskell, as 90% of functional stuff does.